### PR TITLE
[5.2] Use BaseCollection for non-model returns

### PR DESCRIPTION
**Pull request by rkgrep**
Rebased #9191 to 5.2

### CODE REVIEW

**Comment by JosephSilber** [comment]
You can replace all `(new BaseCollection($this->items))` with `$this->toBase()`.

**Comment by JosephSilber** [comment]
I just wish there was a better way to do this. Too bad PHP doesn't let you intercept method calls before they're dispatched.

**Comment by rkgrep** [comment]
@JosephSilber another way is to implicitly return Collection instance in the base collection methods. But I don't remember why I chose this solution previous time.

**Comment by JosephSilber** [comment]
You wouldn't want to do that. If you extend the collection class yourself, you'd want all these methods to return an instance of your collection.

File: tests/Database/DatabaseEloquentCollectionTest.php [hunk 1]
```diff
@@ -229,6 +230,18 @@ public function testWithHiddenSetsHiddenOnEntireCollection()
...
+    public function testNonModelRelatedMethods()
+    {
+        $a = new Collection([['foo' => 'bar'], ['foo' => 'baz']]);
+        $b = new Collection(['a', 'b', 'c']);
+        $this->assertInstanceOf(BaseCollection::class, $a->pluck('foo'));
+        $this->assertInstanceOf(BaseCollection::class, $a->keys());
+        $this->assertInstanceOf(BaseCollection::class, $a->collapse());
+        $this->assertInstanceOf(BaseCollection::class, $a->flatten());
+        $this->assertInstanceOf(BaseCollection::class, $a->zip(['a', 'b'], ['c', 'd']));
+        $this->assertInstanceOf(BaseCollection::class, $b->flip());
```
Comment (JosephSilber) [inline tests/Database/DatabaseEloquentCollectionTest.php]: These don't really test what you want.

An Eloquent collection is an instance of the base collection.

**Pull request merged**